Introduce C library for Pkl

bioball opened the pull request:
This introduces native C bindings for Pkl.

This uses native-image to generate a C library for Pkl. This generated library from native-image is wrapped with our own library, in `pkl.h`.

This produces a static and a dynamic library for each os/arch variant that Pkl currently supports.

## Testing

This runs two types of tests:
1. Message passing tests by calling into libpkl from Java using JNA (binds only to dynamic lib on all platforms)
2. Test executables (smoke test, link against both dynamic and static libs). The test executable calls some methods and exits with 0 if all succeed.

## Distribution

This distributes an archive for each target (zip for Windows, tar.gz for macOS/Linux).

Each archive looks something like this:

```
.
└── libpkl-0.33.0-macos-aarch64
    ├── include
    │   └── pkl.h
    ├── lib
    │   ├── libpkl.a
    │   ├── libpkl.dylib
    │   └── pkgconfig
    │       └── libpkl.pc
    ├── LICENSE.txt
    ├── README.md
    └── THIRD-PARTY-NOTICES.txt
```

## Building libpkl

The way we build this library is quite roundabout. At the core, we use native-image to produce a shared library. However, by default, we're quite restricted in the type of library that we can produce:
1. The header file is generated by GraalVM
2. The header exports types from GraalVM (e.g. graal_isolate)

We have a couple requirements here:
1. We want to hide GraalVM from our own API
2. We want to produce both static and shared libs, so that our users can choose whether they want to embed Pkl directly inside their own tool or not.

To do this, we do the following:
1. GraalVM builds an _internal_ library, called `libpkl_internal`
2. We give GraalVM a _shimmed_ C compiler. This shimmed compiler also builds a static library of our internal lib
3. We build our final shared and dynamic libs using the static library


TODO, not in this PR:
* generate debug builds and debug symbols (.dSYM on macOS, DWARF on Windows, etc)
* add documentation; this should have its own page on the website

[native-libpkl] [native-pkl-cli]

Supersedes #1136

colinmarc commented:
This would be an amazing addition for those of us wanting to use pkl from rust!

z-jxy commented:
Curious if there were any changes planned for the C API? 

I have an implementation against [pkl.h (46db74a)](https://github.com/apple/pkl/blob/46db74a14199265378641c89bb03521081ed7505/libpkl/src/main/c/pkl.h) from a few months back and was planning on doing a safe wrapper around the bindings, but don't want to get too far ahead if things might change before its released

bioball committed:
This introduces native C bindings for Pkl.

This uses native-image to generate a C library for Pkl.
This generated library from native-image is wrapped with our own library,
in `pkl.h`.

This produces a static and a dynamic library for each os/arch variant
that Pkl currently supports.

Co-authored-by: Kushal Pisavadia <kushal.p@apple.com>

bioball commented:
@z-jxy I don't anticipate too many more changes to the `pkl.h` API, although it's always possible that something will change before this is merged

KushalP commented:
@bioball can you merge this into a feature branch please? I'll then open PRs on top.

bioball commented on libpkl/src/main/files/README.md:
```diff
@@ -0,0 +1,5 @@
+
+libpkl is a native C library for calling into Pkl via its message passing API.
+
+For full usage instructions, visit https://pkl-lang.org/main/current/libpkl/index.html
```
This URL doesn't exist yet; I'll submit a follow-up PR to add docs for libpkl.

bioball commented on libpkl/src/main/files/THIRD-PARTY-NOTICES.txt:
```diff
@@ -0,0 +1,437 @@
+The Apache License
+Copyright © 2016 Sadayuki Furuhashi, Muga Nishizawa, Taro L. Saito, Mitsunori Komatsu, Ozawa Tsuyoshi
+Licensed under the Apache License, Version 2.0 (the “License”); you may not use this file except in compliance with the License.You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an “AS IS” BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License. 
```
This gets included in the archives that we create for these built libraries

HT154 commented on libpkl/src/main/c/include/pkl.h:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Returns the version of Pkl in use.
+ *
+ * @return a string with the version information.
```
I'm assuming this is null-terminated because there's no explicit length anywhere:
```suggestion
 * @return a null-terminated string with the version information.
```

bioball replied:
Maybe a little overly clear? Are there any C strings that are not null-terminated?

HT154 replied:
Hmm, I guess... Don't think it hurts to specify

HT154 commented on libpkl/src/main/c/pkl.c:
```diff
@@ -0,0 +1,229 @@
+	return 0;
+}
+
+int pkl_send_message(pkl_exec_t *pexec, int length, char *message,
```
Any reason not to express this in our public C API?

```suggestion
int pkl_send_message(pkl_exec_t *pexec, unsigned int length, char *message,
```

Might be good to have a matching assert in `NativeInputStream` and/or `NativeTransport` as well.

bioball replied:
Good idea RE: `unsigned`.

HT154 commented on libpkl/src/main/java/org/pkl/libpkl/LibPklInternal.java:
```diff
@@ -0,0 +1,126 @@
+   *     stopped and no further responses will ever be produced.
+   */
+  @CEntryPoint(name = "pkl_internal_poll_response")
+  public static int pklInternalPollResponse(IsolateThread thread, CCharPointerPointer outMessage) {
```
Is `thread` actually necessary here? It appears unused.

bioball replied:
Yup, this is a required parameter for `@CEntryPoint` methods (see https://www.graalvm.org/sdk/javadoc/org/graalvm/nativeimage/c/function/CEntryPoint.html)

HT154 commented on libpkl/src/main/c/pkl.c:
```diff
@@ -0,0 +1,348 @@
+  if (pexec->queue_thread == NULL) {
+    if (error != NULL) {
+      error->message = "Failed to create response dispatch thread";
+    }
+    return -1;
+  }
+#else
+	if (pthread_create(&pexec->queue_thread, NULL, pkl_dispatch_worker, pexec) != 0) {
+		if (error != NULL) {
+			error->message = "Failed to create response dispatch thread";
+		}
+		return -1;
+	}
+#endif
+	return 0;
```
Nit: can be deduped a little more
```suggestion
  pexec->queue_thread = CreateThread(NULL, 0, pkl_dispatch_worker, pexec, 0, NULL);
  if (pexec->queue_thread == NULL) {
#else
	if (pthread_create(&pexec->queue_thread, NULL, pkl_dispatch_worker, pexec) != 0) {
#endif
		if (error != NULL) {
			error->message = "Failed to create response dispatch thread";
		}
		return -1;
	}
	return 0;
```

bioball commented on build-logic/src/main/kotlin/NativeImageBuild.kt:
```diff
@@ -111,6 +175,7 @@ abstract class NativeImageBuild : DefaultTask() {
       workingDir(outputDir)
 
       args = buildList {
+        add("--color=always")
```
Quality of life improvement: make native-image emit colored output in our build logs

bioball commented on libpkl/libpkl.gradle.kts:
```diff
@@ -0,0 +1,529 @@
+      "// ",
+    )
+    target("src/*/c/*.c", "src/*/c/*.h")
+    eclipseCdt(libs.versions.eclipseCdtFormat.get())
```
The more popular formatter is clang-format. However, spotless  isn't very good at managing this (see https://github.com/diffplug/spotless/issues/673).

Maybe once that's resolved, we can switch to clang-format.

HT154 commented on libpkl/src/nativeTest/c/test_pkl.c:
```diff
@@ -0,0 +1,75 @@
+#include "pkl.h"
+
+// Callback for receiving messages
+void test_message_handler(int length, char *message, void *userData) {
```
Interesting that this only failed on macOS!
```suggestion
void test_message_handler(unsigned int length, char *message, void *userData) {
```

HT154 reviewed (approved):
Awesome!! Approving to unblock, just a few small things.

bioball committed:
Update test_pkl.c

Co-authored-by: Jen Basch <jbasch94@gmail.com>

KushalP commented on libpkl/libpkl.gradle.kts:
```diff
@@ -0,0 +1,529 @@
+    if (buildInfo.os.isMacOS) {
+      frameworks.addAll("Foundation", "CoreServices")
+      linkerFlags.addAll("-current_version", project.version.toString())
+      linkerFlags.addAll("-compatibility_version", "0.1.0")
```
Should this be hard-coded?

bioball replied:
Good call-out, I meant to call this as an open question.

Here, I'm treating "current_version" as Pkl's own version, and "compatibility_version" as the ABI compatibility version.

So, we wouldn't bump "compatibility_version" unless pkl.h has a breaking change.

Thoughts?

KushalP replied:
Yeah, makes sense. We need a paragraph or so on this as docs. Can be a follow up as long as it's tracked.

bioball replied:
Yeah; docs are a TODO

KushalP commented on libpkl/libpkl.gradle.kts:
```diff
@@ -0,0 +1,529 @@
+
+    if (buildInfo.os.isMacOS) {
+      frameworks.addAll("Foundation", "CoreServices")
+      linkerFlags.addAll("-current_version", project.version.toString())
```
```suggestion
      // macOS `ld` only accepts a numeric X[.Y[.Z]] here, so strip any pre-release/build suffix (e.g. "0.33.0-SNAPSHOT" -> "0.33.0").
      val currentVersion = project.version.toString().substringBefore("-").substringBefore("+")
      linkerFlags.addAll("-current_version", currentVersion)
```

bioball replied:
`project.version.toString()` does give just the X.Y.Z form (see our gradle.properties)

KushalP replied:
Interesting. This gave me the snapshot suffix locally.

bioball replied:
Are you seeing an error when you `./gradlew libpkl:buildNative`?

KushalP commented on libpkl/src/nativeTest/c/test_pkl.c:
```diff
@@ -0,0 +1,75 @@
+	result = pkl_close(NULL, &err);
+	assert(result == -1);
+	printf("✓ Error handling works\n");
+}
```
I think we're missing a test here:

```c
// A zero-length message decodes to nothing and must be reported as an error, not silently accepted.
void test_empty_message() {
	pkl_error_t err = { 0 };
	pkl_exec_t *exec = NULL;
	assert(pkl_init(test_message_handler, NULL, &exec, &err) == 0);

	char dummy = 0;
	int result = pkl_send_message(exec, 0, &dummy, &err);
	assert(result != 0);

	assert(pkl_close(exec, &err) == 0);
	printf("✓ Empty message rejected\n");
}
```

bioball replied:
Great catch, this was actually a bug!

KushalP commented on libpkl/src/nativeTest/kotlin/org/pkl/libpkl/LibPklTest.kt:
```diff
@@ -0,0 +1,52 @@
+    assertThat(result2).isNotEqualTo(0)
+    assertThat(error.message).isEqualTo("Invalid encoding: Malformed message header.")
+    assertThat(LibPklJNA.INSTANCE.pkl_close(execRef.value, error)).isEqualTo(0)
+  }
```
Similar to this, another test we're missing:

```kotlin
  @Test
  fun testEmptyMessageIsRejected() {
    val execRef = PointerByReference()
    val error = LibPklJNA.PklError()
    assertThat(LibPklJNA.INSTANCE.pkl_init(noopHandler, Pointer.NULL, execRef, error))
      .`as` { "Failed to call pkl_init: ${error.message}" }
      .isEqualTo(0)
    try {
      val result = LibPklJNA.INSTANCE.pkl_send_message(execRef.value, 0, byteArrayOf(0), error)
      assertThat(result).isNotEqualTo(0)
      assertThat(error.message).contains("Unexpected end of input")
    } finally {
      assertThat(LibPklJNA.INSTANCE.pkl_close(execRef.value, error)).isEqualTo(0)
    }
  }
```

KushalP commented on libpkl/src/main/java/org/pkl/libpkl/NativeTransport.java:
```diff
@@ -0,0 +1,73 @@
+    } catch (ProtocolException e) {
+      System.err.println("Received unexpected ProtocolException when encoding a message, aborting");
+      e.printStackTrace(System.err);
+      System.exit(1);
+    }
+  }
+
+  public void sendMessage(int length, CCharPointer ptr) throws ProtocolException {
+    try (var is = new NativeInputStream(length, ptr);
+        var unpacker = MessagePack.newDefaultUnpacker(is)) {
+      var message = new ServerMessagePackDecoder(unpacker).decode();
+      assert message != null;
+      accept(message);
+    } catch (IOException e) {
+      // impossible; no IO happens during unpacking
+      throw PklBugException.unreachableCode();
+    }
+  }
```
I think we split out the different kinds of exception here, as `accept` can lead to legit I/O exceptions we need to provide as feedback, whereas the decoder can be swallowed.

```suggestion
  public void sendMessage(int length, CCharPointer ptr) throws ProtocolException {
    Message message;
    try (var is = new NativeInputStream(length, ptr);
        var unpacker = MessagePack.newDefaultUnpacker(is)) {
      message = new ServerMessagePackDecoder(unpacker).decode();
    } catch (IOException e) {
      // impossible; NativeInputStream performs no IO
      throw PklBugException.unreachableCode();
    }
    if (message == null) {
      throw new ProtocolException("Unexpected end of input while decoding message.");
    }
    try {
      accept(message);
    } catch (IOException e) {
      throw new ProtocolException("I/O error while handling message.", e);
    }
  }
```

bioball replied:
Looking at this code, I'm actually not sure why `accept()` is marked as throwing `IOException`; it doesn't seem like this is actually thrown by an implementation. I'm changing the signature for `accept()` instead.

KushalP commented on libpkl/src/main/c/pkl.exported_symbols:
```diff
@@ -0,0 +1,4 @@
+_pkl_init
+_pkl_send_message
+_pkl_close
+_pkl_version
```
line ending?

KushalP commented on libpkl/src/main/c/pkl.map:
```diff
@@ -0,0 +1,9 @@
+    pkl_version;
+  local:
+    *;
+};
```
line ending?

KushalP commented on libpkl/src/main/c/pkl.c:
```diff
@@ -0,0 +1,358 @@
+#endif
+};
+
+static int pkl_is_initialized = 0;
```
How many instances of Pkl/GraalVM are allowed to run per-process? Because this could require us to implement a better "singleton" pattern.

bioball replied:
We only expect one instance. Calling `pkl_init` multiple times (without `pkl_close`) results in an error.

Note: you can create multiple individual evaluators; the executor here is to arbitrate messages through the message passing API. Not sure why you'd want multiple executors for this.

stackoverflow commented on libpkl/libpkl.gradle.kts:
```diff
@@ -0,0 +1,529 @@
+    onlyIf { isWindows }
+    dependsOn(buildSharedLibrary, buildStaticLibrary, processFiles)
+    val baseName = "libpkl-${buildInfo.pklVersionNonUnique}-${buildInfo.targetMachine.targetName}"
+    archiveFileName = "${baseName}.tar.gz"
```
Shouldn't this be `.zip` for windows?

bioball replied:
Good catch!

stackoverflow commented on libpkl/libpkl.gradle.kts:
```diff
@@ -0,0 +1,529 @@
+val createArtifacts =
+  tasks.register("createArtifacts") {
+    dependsOn(buildSharedLibrary, buildStaticLibrary, tasks.processResources)
+    doLast { copy { from(layout.buildDirectory.file("resources/")) } }
```
Don't you need an `into` if you use a `from`? Is this task called somewhere?

bioball replied:
Ah, good catch! This was leftover code.

stackoverflow commented on libpkl/scripts/build_unix.sh:
```diff
@@ -0,0 +1,114 @@
+# and additionally generates a static library.
+#
+# Use with --native-compiler-path=${pathToThisScript}
+
```
Any reason to not use `set -e`?

stackoverflow commented on libpkl/scripts/build_unix.sh:
```diff
@@ -0,0 +1,114 @@
+CC_ARGS=$*
+echo "cc $CC_ARGS" >> "${LOG_FILE}"
+# shellcheck disable=SC2086
+cc $CC_ARGS
```
```suggestion
CC_ARGS=( "$@" )
echo "cc ${CC_ARGS[*]@Q}" >> "${LOG_FILE}"
cc "${CC_ARGS[@]}"
```
`cc $CC_ARGS` is dangerous and might lead to a wrong command depending on the input.

stackoverflow commented on libpkl/src/main/c/pkl.c:
```diff
@@ -0,0 +1,358 @@
+	if (lock_response != 0) {
+		return lock_response;
+	}
+	char *errormessage;
```
```suggestion
	char *errormessage = NULL;
```
Better a null pointer than garbage, if `pkl_internal_send_message`, for some reason, doesn't set it.

stackoverflow reviewed (approved):
Great work! Some small comments.

bioball committed:
Address PR comments

bioball committed:
Adjust docs
